contracts/farm/LiquidityMiningV2.sol

```diff
@@ -925,4 +934,86 @@ contract LiquidityMiningV2 is ILiquidityMining, LiquidityMiningStorageV2 {
 		uint256 poolId = _getPoolId(_poolToken);
 		return userInfoMap[poolId][_user].amount;
 	}
+
+	/**
+	 * @notice read all pools from liquidity mining V1 contract and add them
+	 */
+	function migratePools() external onlyAuthorized {
+		(
+			address[] memory _poolToken,
+			uint96[] memory _allocationPoints,
+			uint256[] memory _lastRewardBlock,
+			uint256[] memory _accumulatedRewardPerShare
+		) = liquidityMiningV1.getPoolInfoListArray();
+
+		require(!migrationFinished, "Migration has already ended");
+		require(_poolToken.length == _allocationPoints.length, "Arrays mismatch");
+		require(_poolToken.length == _lastRewardBlock.length, "Arrays mismatch");
+
+		liquidityMiningV1.finishMigrationGracePeriod();
+		for (uint256 i = 0; i < _poolToken.length; i++) {
+			address poolToken = _poolToken[i];
+			uint96[] memory allocationPoints = new uint96[](1);
+			allocationPoints[0] = _allocationPoints[i];
+			uint256 lastRewardBlock = _lastRewardBlock[i];
+			uint256 accumulatedRewardPerShare = _accumulatedRewardPerShare[i];
+			address[] memory SOVAddress = new address[](1);
+			SOVAddress[0] = address(SOV);
+			//add will revert if poolToken is invalid or if it was already added
+			add(poolToken, SOVAddress, allocationPoints, false);
+
+			uint256 poolId = _getPoolId(poolToken);
+			PoolInfoRewardToken storage poolInfoRewardToken = poolInfoRewardTokensMap[poolId][address(SOV)];
+			//add pool function put lastRewardBlock with current block number value, so we need to retrieve the original
+			poolInfoRewardToken.lastRewardBlock = lastRewardBlock;
+			poolInfoRewardToken.accumulatedRewardPerShare = accumulatedRewardPerShare;
+		}
+		RewardToken storage rewardToken = rewardTokensMap[address(SOV)];
+		rewardToken.startBlock = liquidityMiningV1.getStartBlock();
+		rewardToken.totalUsersBalance = liquidityMiningV1.getTotalUsersBalance();
+	}
+
+	/**
+	 * @notice read all users of all the pools from liquidity mining V1 contract and copy their info
+	 * @param _users a list of users to be copied
+	 */
+	function migrateUsers(address[] calldata _users) external onlyAuthorized {
+		require(!migrationFinished, "Migration has already ended");
+		liquidityMiningV1.finishMigrationGracePeriod();
+		for (uint256 i = 0; i < _users.length; i++) {
+			(uint256[] memory _amount, uint256[] memory _rewardDebt, uint256[] memory _accumulatedReward) =
+				liquidityMiningV1.getUserInfoListArray(_users[i]);
+
+			require(_amount.length == _rewardDebt.length, "Arrays mismatch");
+			require(_amount.length == _accumulatedReward.length, "Arrays mismatch");
+
+			address user = _users[i];
+
+			require(userMigrated[user] == false, "User already migrated");
```

Comment:
do we want to make the whole transaction fail or just skip this one? :thinking:

Reply:
yes, maybe it is better only skip